docs/blog/2020/CPP-guide.md

```diff
@@ -0,0 +1,275 @@
+disqus: emacs-lsp
+---
+
+# Configuring `Emacs` a `C/C++` IDE
+
+In this guide, I will show you how to configure `lsp-mode` and `dap-mode` for C/C++ development, using `GNU Emacs` as an example code base. At the end of this tutorial, Emacs should provide you with all the language-aware editor features you'd expect from a modern C++ IDE, such as
+
+* Code completion...
+environment. The guide is mainly focused on `lsp-mode`/`dap-mode` and the
+features that they have. We will use `GNU Emacs` as a project in order to illustrate each of
+the steps to get `C/C++` project up and running.
+
+Some of the main features that we can expect from C++ IDE:
+
+* Code completion
+* Real-time syntax checking
+* Code navigation (references/definitions/implementations/symbol search)
+* Visual IDE debugger
+* Breadcrumb on headerline
+
+For all `lsp-mode` features, check [here](https://emacs-lsp.github.io/lsp-mode/page/main-features/).
+
+# Obtaining language server
+
+`lsp-mode` is a client-server application with `Emacs` acting as the client.
+In order for it to work, you have to install a separate _language server_ that understands the specifics of your target language. As of late 2020, there are 2 production-ready  language servers for `C/C++` and `Objective-C`, [ccls](https://github.com/MaskRay/ccls) and [clangd](https://clangd.llvm.org/installation.html), with `ccls`'s precursor `cquery` being deprecated.
+used by `Emacs` as provider for language specific features(completion,
+diagnostics, navigation, etc). For some of the language servers `lsp-mode`
+ships with automatic installation scripts but there is no such script for
+installing `C/C++` language server due to their overall complexity. As of end of
+2020 there are 2 `C/C++` language servers - `ccls` and `clangd` with `cquery`
+being deprecated. For this guide we will be using `clangd` but most of the steps here
+apply to `ccls` as well. `lsp-mode` is pre-configured and it will be able to find
+the language server automatically when it is installed on the host machine and
+it is present on the path. In case `clangd` is not present on the path
+`clangd` can be installed following their [Getting Started](https://clangd.llvm.org/installation.html "Getting Started")
+instructions.
+
+_Note:_ for `Debian` based system use https://apt.llvm.org/
+
+## Project setup
+
+To understand your source code, `clangd` needs to know your build flags. (This is
+just a fact of life in C++, source files are not self-contained).
+
+By default, `clangd` will assume your code is built as `clang` `some_file.cc`, and
+you’ll probably get spurious errors about missing `#included` files, etc. There
+are a couple of ways to fix this. You may read about them in [installation](https://clangd.llvm.org/installation.html "clangd installation")
+
+In order to generate `compile_commands.json` for `Emacs` itself I have found
+that [Bear](https://github.com/rizsotto/Bear) works fine.
+
+``` bash
+git clone https://github.com/rizsotto/Bear
```

Comment:
`Bear` can be installed via `apt` too `sudo apt install bear`

Reply:
thanks, updated.